src/main/kotlin/com/exactpro/th2/common/microservice/IApplicationFactory.kt

```diff
@@ -0,0 +1,162 @@
+/*
+ * Copyright 2023 Exactpro (Exactpro Systems Limited)
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.exactpro.th2.common.microservice
+
+import com.exactpro.th2.common.metrics.MetricMonitor
+import com.exactpro.th2.common.metrics.registerLiveness
+import com.exactpro.th2.common.metrics.registerReadiness
+import com.exactpro.th2.common.schema.factory.CommonFactory
+import mu.KotlinLogging
+import java.util.*
+import java.util.concurrent.ConcurrentLinkedDeque
+import java.util.concurrent.locks.Condition
+import java.util.concurrent.locks.ReentrantLock
+import kotlin.concurrent.thread
+import kotlin.concurrent.withLock
+import kotlin.system.exitProcess
+
+/**
+ * Factory can share closable resources between application. They should be closed on [IApplicationFactory.close] method call
+ */
+interface IApplicationFactory : AutoCloseable {
+    /**
+     * Creates onetime application.
+     * If you need restart of reconfigure application, close old instance and create new one.
+     */
+    fun createApplication(context: ApplicationContext): IApplication
+
+    /**
+     * Close resources shared between created application.
+     */
+    override fun close() {}
+
+    companion object {
+        private val K_LOGGER = KotlinLogging.logger {}
+        private const val APPLICATION_FACTORY_SYSTEM_PROPERTY = "th2.microservice.application-factory"
+        private const val MONITOR_NAME = "microservice_main"
+        @JvmStatic
+        fun run(factory: IApplicationFactory?, vararg args: String) {
```

Comment:
We can set the `null` default value to `factory` and add `@JvmOverloads` annotations to have a method that accepts only varargs